### PR TITLE
Remove duplicate run of post-command hook.

### LINKED ISSUE #248: Fix bugs in PRE/POST hooks
* If a post hook command fails, this triggers a second run of the post hooks.
 * When sanity checks run,  post hooks are run, but not pre hooks -- suggest running neither.

### PR DESCRIPTION
**Pull request by snim2**
Not yet ready for review.

Fixes #248

### CODE REVIEW

File: krun.py [hunk 1]
```diff
@@ -210,7 +210,6 @@ def main(parser):
     except util.FatalKrunError as e:
         subject = "Fatal Krun Exception"
         mailer.send(subject, e.args[0], bypass_limiter=True)
-        util.run_shell_cmd_list(config.POST_EXECUTION_CMDS)
```
Comment (snim2) [inline krun.py]: This is the source of the duplicate call when a post-command fails, and `util.run_shell_command` calls `fatal()`. So, do you want to introduce a new exception type? `JustAFleshWoundKrunError`?

**Comment by ltratt** [comment]
I wonder if we're being too clever for our own good here. I think the simplest solution might in fact be the right one. Here are the rules I think we have to observe:
 * If the pre commands are run, the post commands *must* be run
 * If *any* of the pre commands are run, the post commands *must* be run
I don't think we've articulated that second point before, but it now seems obvious to me. If, for example, an early pre command brings the network down, and then a later pre command fails, we still want the post commands to be executed (to bring the network back up).

On that basis, I presume that all we need is:
```python
try:
  exec_pre_cmds()
  <everything else>
finally:
  exec_post_cmds()
```

**Comment by vext01** [comment]
In addition, the pre-and post commands need to happen before and after each process execution even if there is no error. As it stands, in reboot mode, the finally would not be reached if no error occurred because Krun reboots the system.

**Comment by ltratt** [comment]
I thought it raised an exception to indicate that it wanted to reboot? If not... can we make it raise an exception :)

Then it can be something like:

```python
reboot = False
try:
  exec_pre_cmds()
  <everything else>
except RebootException:
  reboot = True
finally:
  exec_post_cmds()

if reboot:
  self.do_reboot()
```

**Comment by vext01** [comment]
That would work for reboot mode, but not for no-reboot mode.

In another PR, I suggested removed no-reboot mode might not be a bad idea.

**Comment by ltratt** [comment]
OK, so let's deprecate no-reboot mode? Then this PR could simply put in an error saying "no reboot mode no longer supported" and a separate PR can actually do the ripping out?

**Comment by vext01** [comment]
Sounds good.

Makes sense @snim2?

**Comment by vext01** [comment]
If there is one thing I have learned from Krun, it is, don't add so many knobs and whistles that it becomes hard to test. This is what I inadvertently did with Krun.

**Comment by ltratt** [comment]
Amen to simplicity.

**Comment by snim2** [comment]
I think it might be simpler to do both things in one PR, TBH

**Comment by ltratt** [comment]
I'm not a big of big PRs. The ripping-out PR will quite probably obscure the (very important) logic this PR needs to concentrate on.

**Comment by vext01** [comment]
Let's also add an exception handler for *all* other kinds of exception, so that we will get an email if (e.g.) and assertion fires.

**Comment by snim2** [comment]
I am going to bring this up to date with master and force push, since we have made substantial changes.

**Comment by ltratt** [comment]
OK.

**Comment by vext01** [comment]
So where do we stand with this?

**Comment by ltratt** [comment]
I don't think anything's changed relative to my suggestions above?

**Comment by snim2** [comment]
I've rebased and I'll push when I have some new commits.

**Comment by snim2** [comment]
I've had a first go at this. Note that if a pre- or post-execution command fails, Krun now stops. I'm not sure if that is what was intended in the discussion above?

File: krun/scheduler.py [hunk 2]
```diff
@@ -345,56 +345,72 @@ def run(self):
...
-        )
+        # Run the pre-exec commands, the benchmark and the post-exec commands.
+        # These are wrapped in a try/except, so that the post-exec commands
+        # are always executed, even if an exception has occurred. We only
+        # reboot /after/ the post-exec commands have completed.
+        try:
+            # Run the user's pre-process-execution commands We can't put an ETA
+            # estimate in the environment for the pre-commands as we have not
+            # (and should not) load the results file into memory yet.
+            util.run_shell_cmd_list(
```
Comment (ltratt) [inline krun/scheduler.py]: Should be on one line.

File: krun/scheduler.py [hunk 3]
```diff
@@ -345,56 +345,72 @@ def run(self):
...
+        except Exception:
+            error_info = sys.exc_info()
+            lines = ["Krun error during benchmark execution: %s\n" % str(error_info[1])]
+            for frame in traceback.format_tb(error_info[2]):
+                lines.append(frame)
+            msg = "".join(lines)
+            util.log_and_mail(self.mailer, debug,
+                              "Krun error during benchmark execution",
+                              msg, bypass_limiter=False)
+
```
Comment (ltratt) [inline krun/scheduler.py]: Add `sys.exit(1)` here?

File: krun.py [hunk 4]
```diff
@@ -192,7 +192,6 @@ def main(parser):
     try:
         inner_main(mailer, on_first_invocation, config, args)
     except util.FatalKrunError as e:
```
Comment (vext01) [inline krun.py]: Should this be a generic except?

Comment (vext01) [inline krun.py]: (I'm not sure)

File: krun/scheduler.py (hunk 3 above)
Comment (vext01) [inline krun/scheduler.py]: extra line?

File: krun/scheduler.py [hunk 5]
```diff
@@ -345,56 +345,72 @@ def run(self):
...
+            if self.platform.hardware_reboots:
+                # Add time taken to wait for system to come up if we are in
+                # hardware-reboot mode.
+                eta_info += STARTUP_WAIT_SECONDS
+            self.add_eta_info(job.key, eta_info)
+
+            # We dump the json after each process exec so we can monitor the
+            # JSON file mid-run. It is overwritten each time.
+            self.manifest.update(flag)
+
```
Comment (snim2) [inline krun/scheduler.py]: I was following the structure of your code here, but can lose a few lines

File: krun/scheduler.py (hunk 3 above)
Comment (snim2) [inline krun/scheduler.py]: Surely the post-exec commands in the `finally` won't happen if we `sys.exit(1)` here?

Comment (ltratt) [inline krun/scheduler.py]: I think `sys.exit` itself raises an exception? [It's been a while, so I don't remember fully.]

**Comment by ltratt** [comment]
I'm fine with just reraising, so that's fine. Can we lose the blank lines before the `except` and `finally` please?

**Comment by snim2** [comment]
Ready to squash?

**Comment by ltratt** [comment]
Let's rebase against master, squash, and then I can merge.

**Pull request merged**